### ISSUE TITLE
Dynamic provider configuration assignment

### ISSUE THREAD

**Issue by RafPe** [opentofu/opentofu#300]
Suggest an existing issue in legacy Terraform to fix in OpenTF. 

To increase automation we struggle many times when configuring providers. It would be great to finally be able to configure them dynamically or at least to be able to dynamically reference them instead of having all of that statically typed

---

Implementation PR: https://github.com/opentofu/opentofu/pull/2054

**Comment by roni-frantchi** [comment]
Hey @RafPe  thanks for submitting!

To help us maintain a clear separation between opentf and hashicorp's offerings, we're asking that people describe issues that are in other repositories rather than linking those directly.
I've thus scrubbed out any links to said issues/PRs.

If there's any more context or description to the problem you think would be good to share and add in please do.  

Thanks!

**Comment by RafPe** [comment]
@roni-frantchi Sure - I understand. Let me explain more in detail what I mean here. 

## Situation now : we define our providers more less in a static way ( there is a level where we can use variables ) and then reference them completely static (no vars, no dynamics etc ) 

```
provider "aws" {
  region = "eu-central-1"
  alias  = "x1"
}

provider "aws" {
  region = "eu-west-1"
  alias  = "x2"
}



resource "aws_resource" "x1" {
  provider = aws.x1
}

resource "aws_resource" "x2" {
  provider = aws.x2
}
```

## Scenario 1- dynamically referencing providers: Having the above providers, we could either have something in form of `singleton` object called `providers` to which one we could reference via alias the pointer to the provider we defined. 

```
provider "aws" {
  region = "eu-central-1"
  alias  = "x1"
}

provider "aws" {
  region = "eu-west-1"
  alias  = "x2"
}

resource "aws_resource" "x1" {
  provider = providers.aws["x1"]
}

resource "aws_resource" "x2" {
  provider = providers.aws["x2"]
}

```

## Scenario 2- dynamically referencing providers with for_each/count: Having the same above providers, we could either have something in form of `singleton` object called `providers` to which one we could reference via alias the pointer to the provider we defined but also support for_each/count on the providers level to create them dynamically 

```
locals {
  aws_accounts = [
    { "aws_account_id": "123456789012", "foo_value": "foo",    "bar_value": "bar"    },
    { "aws_account_id": "987654321098", "foo_value": "foofoo", "bar_value": "barbar" },
    ]
}

## Here's the proposed magic... `provider.for_each`
provider "aws" {
  for_each = local.aws_accounts
  alias = each.value.aws_account_id

  assume_role {
    role_arn    = "arn:aws:iam::${each.value.aws_account_id}:role/TerraformAccessRole"
  }
}

## Here is the magic of referencing them in our resources 

resource "aws_resource" "x1" {
  for_each = local.aws_accounts

  provider = providers.aws[each.value.aws_account_id]
}

```

I believe this approach ( especially in bigger organisations ) would allow for more automation and maintaining the logic within the tool instead of leveraging external templating mechanism

Please let me know if the above makes sense :)

**Comment by jwenz723** [comment]
In my situation I have multiple AWS accounts (dev, test, prod). All accounts use the us-west-2 region, but the prod account also uses a few other regions. 

So in my case I really need one provider enabled when deploying to dev and test and multiple providers enabled when deploying to prod. I am able to work around this currently by just including all regional providers in all accounts and I just don’t use the extra providers during a dev or test deploy. 

This work around no longer works though when you have a module that needs to be deployed to multiple AWS partitions. For example standard and govcloud. A govcloud provider will not work when deploying to a standard partition and a standard provider will not work when deploying to govcloud. In this situation I am able to work around this by using some custom code outside of opentofu to generate the provider config before running opentofu. 

It would be nice to just have the ability to do loops and conditionals on providers.

**Comment by timothyjlaurent** [comment]
This could be OpenTF's killer app ;)

**Comment by RafPe** [comment]
@roni-frantchi as I myself would be open to contributing to this ... would eb anyone else experienced already in the code base that could guide me towards making this possible ?

**Comment by roni-frantchi** [comment]
Hey @RafPe !  

> I myself would be open to contributing to this would eb anyone else experienced already in the code base that could guide me towards making this possible ?

Very much appreciated.  
It'd be harder for us to provide such guidance, as right now our core team focuses on:  
1. Rolling out an alpha in the next couple of days
2. Focusing on a design for a registry to power the stable release

Of course if you're willing to dive in there yourself or assisted by anyone else from the community would love to see such a design from you

**Comment by murarisumit** [comment]
Ref to issue in terraform repo, quite a bit of discussion over there on it: https://github.com/hashicorp/terraform/issues/19932#issuecomment-877438031

**Comment by dbhaigh** [comment]
Issue 19932 has been open since Jan 7th 2019 - it doesn't look like it's been given much love since

Plenty of people wanting this ability (myself included)

This ability would be fantastic, and yes, bring people over to the Open Source side

**Comment by Chandra2614** [comment]
This needs to be fixed ASAP, 3 years are a lot

**Comment by RafPe** [comment]
@cube2222 maybe I do not see the whole technical picture - but would it for example be possible to split it in two working items ? 

1)  to make it accessible as just static array so we can index it into via a key 
2) Think of making it super dynamic with variable interpolations

**Comment by cam72cam** [comment]
I think that makes sense, as `#2` would likely relate to #1042

**Comment by ImIOImI** [comment]
Scenario 2 feels like it'll cover most of the cases I encounter in the wild... but I say without any hesitation that no matter how unergonomic the solution is, I'll figure it out and use it and be happy to have it. 

That being said, would it be possible to configure and pass a provider like:

```
locals {
  my_object = [
    { provider: providers.aws["dev"], "foo_value": "foo",    "bar_value": "bar"    },
    { provider: providers.aws["dev"], "foo_value": "apple",    "bar_value": "pear"    },
    { provider: providers.aws["prd"], "foo_value": "foo",    "bar_value": "bar"    },
    ]
  aws_accounts = {
    dev = {
      aws_account_id": "123456789012"
     },
    prd = { 
      "aws_account_id": "987654321098"
    }
}

provider "aws" {
  for_each = local.aws_accounts
  alias = each.key

  assume_role {
    role_arn    = "arn:aws:iam::${each.value.aws_account_id}:role/TerraformAccessRole"
  }
}

resource "aws_resource" "x1" {
  for_each = local.my_object

  provider = each.value.provider
}
```

**Comment by cam72cam** [comment]
Something worth noting, I think the `provider.` prefix is probably required as part of this work (at least for expressions).  Otherwise your provider names are top level identifiers and can conflict with other things like "local"

**Comment by nitrocode** [comment]
In this case this helps for people in the mean time.

I've used workspaces with a specific tfvars containing a variable for the region as a workaround for creating the same resources in multiple regions or multiple accounts.

`main.tf`
```hcl
variable region {}
variable account {}

provider aws {
  region = var.region

  assume_role {
    role_arn = "arn:aws:iam::${module.account_map.accounts[var.account]}:role/cicd"
  }
}
```

`uw2-dev.tfvars`
```hcl
region = "us-west-2"
account = "dev"
```

Commands
```bash
terraform workspace select uw2-dev
terraform init -var-file uw2-dev.tfvars
```

**Comment by ImIOImI** [comment]
> I've used workspaces with a specific tfvars containing a variable for the region as a workaround for creating the same resources in multiple regions or multiple accounts.
>...

As of now, I think you're right about workspaces being the best answer when needing dynamic providers. I've done something similar like the following in Azure:

```terraform
locals {
  cntxts = {
    defaults = {
      subscription_id         = "00000000-0000-0000-0000-000000000000"
      dynamic_subscription_id = "12345678-0000-0000-0000-000000000000"
      tenant_id               = "00000000-0000-0000-0000-000000000000"
    }
    dev = {
      dynamic_subscription_id = "abcdefgh-0000-0000-0000-000000000000"
    }
    infra = {
      dynamic_subscription_id = "abc123hi-0000-0000-0000-000000000000"
    }
    stg = {
      dynamic_subscription_id = "foobar12-0000-0000-0000-000000000000"
    }
    prd = {
      dynamic_subscription_id = "RunningO-utOf-Gene-ricG-UIDideas1234"
    }
  }

  contexts                = module.contexts.merged
  default_subscription_id = local.contexts.subscription_id
  dynamic_subscription_id = local.contexts.dynamic_subscription_id
  tenant_id               = local.contexts.tenant_id

  # where workspaces are like (dev, infra, stg, prd)
  context = terraform.workspace
}

module "contexts" {
  source = "Invicton-Labs/deepmerge/null"
  maps   = [
    local.cntxts.defaults,
    local.cntxts[local.context],
  ]
}

provider "azurerm" {
  alias = "default"

  subscription_id = local.default_subscription_id
  tenant_id       = local.tenant_id
}

provider "azurerm" {
  alias = "dynamic"

  subscription_id = local.dynamic_subscription_id
  tenant_id       = local.tenant_id
}

```

However, this often forces you into architecture patterns that could be more simply solved with a loop of providers

**Comment by vierkean** [comment]
Hello,
since HashiCorp still supported this function in terraform version 0.11 and no longer does since 0.12, I was looking for an alternative to terraform and came across tofu.
Unfortunately, the current version of tofu does not support it either.
I hope tofu will support it soon and I don't have to rebuild the organization in terraform to use a current version.

`provider "aws" {
    alias = "${var.AccountName}"
    region = "${var.region}"
    assume_role {
        role_arn = "arn:aws:iam::${aws_organizations_account.account.id}:role/someRole"
    }
}`

Kind Regards

**Comment by cam72cam** [comment]
This has been discussed and accepted as defined in https://github.com/opentofu/opentofu/blob/main/rfc/20240513-static-evaluation-providers.md.  We are tentatively aiming for including this in OpenTofu 1.9.0 as an extension to the existing static evaluation work that is present in 1.8.0.

**Comment by tmccombs** [comment]
Something  I would like to see, maybe as a future extension of this, would be to have a way to pass a dynamic number of providers to a module.  So that a module would be able to use a for_each over a set of providers passed in to the module.

I'm not sure what the syntax for that would look like though.

**Comment by cam72cam** [comment]
Tentatively moving the the 1.9.0 milestone as prototyping is currently in progress.

**Comment by cveld** [comment]
@cam72cam will you guys make provider aliases real sub resources in the tree? We use kubernetes providers as a sub from azurerm a lot. Officially this is only supported in the root but we leverage this a bit deeper in the tree. 

I guess we could live with an architecture where a single provider server is running, but with the option to configure aliases at any place in the tree. E.g. any arbitrary subscription id for azurerm or any arbitrary kubernetes config for the kubernetes providers.

**Comment by cam72cam** [comment]
@cveld could you create a new issue and link it to this one?  I want to limit the scope of this to for_each support in modules and resources as an extension on top of where configuration is currently allowed.

**Comment by dbmurphy** [comment]
There is also the challenge where you might have `root_module -> mysql_cluster -> mysql_users -> provider.mysql -> resource.mysql_user` (other providers apply also, but a good example). 

Normally you would want to define a provider at the root_module, however currently in both Terraform and OpenTofu we will get errors if you define a blank config "mysql" provider then inside mysql_cluster module override the config with one pointing to the cluster managed in that module. 

As part of its dynamic nature, supporting configuration overriding in the planning/application phases would be nice. Also, if we could use the stored provider configuration from the state with the module destroy planning to tear down first resources made by that provider, then other resource teams would not need a multiple plan clean-up process like they do today due to these limits.

**Comment by jackfarzan** [comment]
I understand that this feature is in progress, but I'd like to add it'd be great to be able to target a specific provider during an apply, in case one region etc needs to be provisioned first.

**Comment by apparentlymart** [comment]
There is a tricky situation that I don't see discussed in [the current RFC](https://github.com/opentofu/opentofu/blob/0cc4301cd51bf3ef1de4d87edfef2ab81cb678b8/rfc/20240513-static-evaluation-providers.md). :thinking: 

Consider the following configuration:

```hcl
variable "regions" {
  type = set(string)
}

provider "aws" {
  for_each = var.regions

  region = each.value
}

resource "aws_s3_object" "example" {
  for_each = var.regions
  provider = aws[each.value]

  # (the details of this don't matter for what we're discussing here;
  # this could be any resource using the dynamically-expanded
  # provider)
}
```

Now consider the following sequence of commands:

1. `tofu init` (just to install the provider plugin)
2. `tofu apply -var='regions=["us-west-2", "us-east-1"]'`
3. `tofu apply -var='regions=["us-east-1"]'`

After step 2 something like the following would be tracked in the state:

```json
{
  "type": "aws_s3_object",
  "name": "example",
  "provider": "provider[\"registry.opentofu.org/hashicorp/aws\"]",
  "instances": [
    {
      "index_key": "us-east-1",
      "provider": "provider[\"registry.opentofu.org/hashicorp/aws\"].us-east-1"
    },
    {
      "index_key": "us-west-2",
      "provider": "provider[\"registry.opentofu.org/hashicorp/aws\"].us-west-2"
    }
  ]
}
```

This means that when planning in step 3 the current state contains both instances but the desired state only contains the `us-east-1` instance. Tofu would conclude that reconciling this means to destroy `aws_s3_object.example["us-west-2"]`, which involves making a `PlanResourceChange` request (during planning) and an `ApplyResourceChange` request (during the subsequent apply phase) to the provider instance for configuration `aws["us-west-2"]`.

However, in step 3 there is no longer any `aws["us-west-2"]` in the configuration, and so OpenTofu wouldn't know how to configure the provider to plan and apply that "delete" action. I expect that an implementation like what's described in the RFC would generate a dynamic equivalent of the currently-statically-checked message we have today:

```
Error: Provider configuration not present

To work with aws_s3_object.example["us-west-2"] its original provider
configuration at aws["us-west-2"] is required, but it has been removed.

This occurs when a provider configuration is removed while objects
created by that provider still exist in the state. Re-add the provider
configuration to destroy aws_s3_object.example["us-west-2"], after
which you can remove the provider configuration again.
```

However, the advice in the current error message isn't helpful in this case, because without modifying the configuration the only way to "re-add the provider configuration" is to add `us-west-2` back into `var.regions`, which would then cause `aws_s3_object.example["us-west-2"]` to be included in the desired state again.

OpenTofu today generally requires that a provider configuration always live for at least one more plan/apply round than any resource instance that's bound to it, which is impossible to achieve when the presence of both the provider configuration and the resource instance are decided by the same input variable value.

Has something like this come up in the implementation work yet? Is there a clever solution to it that I'm not thinking of yet? :grinning:

**Comment by tmccombs** [comment]
One way of dealing with that is to store information about the configuration in the state itself.

But that still gets tricky, what if the configuration includes credentials? Should those be stored in the state file? What if those credentials have expired?

Another possibility is to have some way to provide configuration for a disposed provider to run rhe cleanup. Maybe something like:

```terraform
removed_provider {
  from = aws["us-west-2"]
  disposed_config  = {
    region = "us-west-2"
  }
}
```

**Comment by air3ijai** [comment]
One of the [proposal](https://github.com/hashicorp/terraform/issues/19932#issuecomment-877438031) was is to
> introduce a new provider attribute, `enabled`

**Comment by kubrickfr** [comment]
A workaround  would be to maintain two different sets, one for provider regions and another one for deployment regions maybe?

```terraform
variable "regions" {
  type = set(string)
}

var "enabled_regions" {
  type =set(string)
  validation {
    condition     = sort(setintersection(regions, enabled_regions)) ==  sort(enabled_regions)
    error_message = "enabled_regions must be a subset of regions"
  }
}

provider "aws" {
  for_each = var.regions
  region = each.value
}

resource "aws_s3_object" "example" {
  for_each = var.enabled_regions
  provider = aws[each.value]
}
```

woud this work?

**Comment by cam72cam** [comment]
@kubrickfr, that's what I had in mind.  I think the main question for me is if this is an acceptable foot-gun if there is comprehensive documentation and refined warning/error messages.

**Comment by ImIOImI** [comment]
I feel like @kubrickfr's solution is acceptable. Personally, I can't imagine changing a variable that "builds" a map of providers then having any other result other than @apparentlymart's listed error. 

As I've mentioned elsewhere, provider errors can be famously hard to diagnose some thing that would be helpful here (and frankly any provider change) would be:
1) the filename/line number where the provider is declared
2) list all the resources the provider in question is tied to
3) in this specific case, mentioning that the provider is being built with a fore_each might make it immediately obvious what's happening along with the above info

**Comment by apparentlymart** [comment]
Maintaining two different sets is indeed a plausible solution, though it is a little risky in that it requires a module author to "plan ahead" and design their module to support removal from the outset even though they might not actually think about this problem until the first time they need to remove a region.

That could be okay, but I'd personally prefer a solution that encourages writing the module "the right way" from the outset, so that an author gets feedback during initial authoring if they've written something that wouldn't allow the provider configuration to outlive the resource instance by one round. I don't know if there _is_ such a solution.

A more general way to think about this problem is that in OpenTofu's current design we need to rely exclusively on the state when something is planned to be deleted. If there were a way to explicitly declare in the configuration how something ought to be deleted -- which provider configuration to use to do it, which destroy-time provisioners to run when it's deleted, and even whether to delete it at all or just "forget" it -- then that might give a stronger building block for talking about this problem.

The `removed` block type is already a way to talk about how to respond to something having been removed, although today it only talks about the removal of entire `resource`/`module` blocks and the only thing it can express is "don't delete this, just forget it". If it were extended to support talking about the removal of individual resource instances ((https://github.com/opentofu/opentofu/issues/1995)) and with the ability to explicitly refer to a provider configuration then it could be used to describe the situation from @kubrickfr's example more explicitly:

```hcl
variable "regions" {
  type = set(string)
}

variable "enabled_regions" {
  type = set(string)
  validation {
    condition     = sort(setintersection(regions, enabled_regions)) ==  sort(enabled_regions)
    error_message = "enabled_regions must be a subset of regions"
  }
}

locals {
  removed_regions = setsubtract(var.regions, var.enabled_regions)
}

provider "aws" {
  for_each = var.regions
  region   = each.value
}

resource "aws_s3_object" "example" {
  for_each = var.enabled_regions
  provider = aws[each.value]
}

removed {
  for_each = local.removed_regions
  from     = aws_s3_object.example[each.value]

  destroy  = true
  provider = aws[each.value]
}
```

An advantage of this being explicit is that then in principle OpenTofu could consider it to be invalid for there to be a resource that's associated with a dynamic-expanded `provider` block without there also being an explicit `removed` block that explains how OpenTofu ought to deal with an instance being removed.

This then moves the problem to initial authoring time rather than being a hazard that you only encounter the first time you try to remove something from the set, though admittedly it also requires writing additional configuration that some might consider redundant boilerplate: it doesn't _actually_ say anything that OpenTofu couldn't infer for itself, but forcing it to be written out means that the module author must decide on _some_ way to describe regions that have been "disabled".

I will say that I'm not feeling 100% certain that this additional boilerplate is actually needed: we could just accept that writing modules with dynamic provider configurations incorrectly is a likely mistake that module authors will learn about at some later date. I just have a worry that a module author might learn about it at a highly inconvenient time, or worse still that the module might be written by someone other than the operator of the current configuration and thus the operator ends up having to fork a shared module just to teach it how to deal with a region being removed. :grimacing:

**Comment by kubrickfr** [comment]
> it requires a module author to "plan ahead" and design their module to support removal from the outset 

Is this so? If a user only has a `regions` variable, wouldn't they be able to create a `provider_regions` variable after the fact just to initialize the providers?

**Comment by kubrickfr** [comment]
I would like to add that today, without dynamic provider, the problem is exactly the same when trying to remove a region: it needs to be done it two steps: first remove the resources, then the provider. I just had the problem in my code-base. So the PR in its current state doesn't make the situation worse IMHO.

**Comment by apparentlymart** [comment]
When I talk about "planning ahead" I am thinking about trickier situations such as where the module is being used by someone other than its author, as a shared module. Situations where someone using a third-party module finds themselves blocked from making progress until that third-party module has been changed in some way cause considerable friction both for the user of the module (who is blocked) and the author of the module (who is suddenly dealing with an apparent emergency they didn't expect).

While it's certainly a tradeoff to be made and not a hard-and-fast rule -- it is still possible we decide that it's okay to have this "trap" to prioritize other goals -- I think we should _prefer_ to have a design that helps module authors write modules that work for the entire lifecycle, including removing things, right from the outset even if they are not yet OpenTofu experts with hard-earned experience with this sort of "trap".

(This is, incidentally, also why OpenTofu does not allow using a `provider` block inside a module that's called with either `count` or `for_each` today, as I mentioned over in https://github.com/opentofu/opentofu/issues/1896#issuecomment-2387465909 .)

**Comment by apparentlymart** [comment]
I'm not currently working directly on the implementation of this feature, so I'll let the folks who are comment on whether the crash you reported in your comment on the PR is expected for the current state of development (it is a draft PR, after all) or whether that's surprising to them.

**Comment by vitaly-dt** [comment]
Will this feature allow us to use the provider block as a dependency?

**Comment by thuck** [comment]
I was testing the `v1.9.0-alpha1` and one feedback is that the error message should include the key somewhere.  
I was testing the datadog provider, and one of my keys was wrong, the error message is:
```
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl,
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
```

As the key is not part of the error, it's very hard to find which one has an issue; it would be nice to have something more like:
```
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl["KEY"],
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
```

**Comment by cam72cam** [comment]
> Will this feature allow us to use the provider block as a dependency?

Not in this feature. Take a look at https://github.com/opentofu/opentofu/pull/2088 for thoughts to expand upon the existing provider configuration.

**Comment by cam72cam** [comment]
@thuck thanks for the feedback, can you post the full error so I can make sure I update that log message?

**Comment by thuck** [comment]
@cam72cam no problem at all.
```
tofu plan

Planning failed. OpenTofu encountered an error while generating this plan.

╷
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl,
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
│ 
╵
╷
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl,
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
│ 
╵
╷
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl,
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
│ 
╵
╷
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl,
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
│ 
╵
╷
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl,
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
│ 
╵
╷
│ Error: 403 Forbidden
│ 
│   with provider["registry.opentofu.org/datadog/datadog"].bl,
│   on main.tf line 36, in provider "datadog":
│   36: provider "datadog" {
│ 
╵
``` 

Here the main.tf that I'm using to test:
```
terraform {
  required_providers {
    datadog = {
      source = "DataDog/datadog"
    }
  }
}

locals {
  datadog = jsondecode(file("dd_keys.json"))
}

provider "datadog" {
  for_each = local.datadog
  alias    = "bl"
  api_url  = each.value.api_url
  api_key  = each.value.api_key
  app_key  = each.value.app_key

}

data "datadog_roles" "foo" {
  for_each = local.datadog
  provider = datadog.bl[each.key]
}
```

**Comment by apparentlymart** [comment]
Hi all!

I realize that yesterday I neglected to explicitly mention here that [v1.9.0-alpha1](https://github.com/opentofu/opentofu/releases/tag/v1.9.0-alpha1) includes an early version of a solution for this feature that we've been working on over in https://github.com/opentofu/opentofu/issues/2123.

If you have the time and interest, we'd love to hear feedback from folks who try this with somewhat-realistic configurations that match whatever problems had caused them to be interested in this issue. **Please don't use this in production, though!**

It's tough to keep track of multiple threads of conversation in a single GitHub issue, and we'd like to keep this issue focused on the use-case rather than on details of the initial solution to it, so if you _do_ try the alpha and have feedback please open a new issue for each feedback item, and then we'll collect links to them in https://github.com/opentofu/opentofu/issues/2123 to make sure we can investigate each one.

Thanks!

**Comment by kylan11** [comment]
I'm attempting to fetch a list of AWS Account IDs from the remote state of a separate OpenTofu configuration, and dynamically generate providers for each on **1.9.0-alpha2**.

Is this outside the scope of the current enhancement?

```hcl
data "terraform_remote_state" "org" {
  backend = "s3"
  config = {
    bucket = var.s3_bucket_name
    key    = "terraform-aws-org/terraform.tfstate"
    region = var.aws_region
  }
}

locals {
  account_ids = values(data.terraform_remote_state.org.outputs.member_account_ids)
}

provider "aws" {
  alias = "child"

  for_each = toset(local.account_ids)

  region = var.aws_region
  assume_role {
    role_arn     = "arn:aws:iam::${each.key}:role/OrganizationAccountAccessRole"
    session_name = "TerraformSession-${each.key}"
  }
}
```


```hcl
tofu plan
╷
│ Error: Dynamic value in static context
│ 
│   on locals.tf line 4, in locals:
│    4:   account_ids = values(data.terraform_remote_state.org.outputs.member_account_ids)
│ 
│ Unable to use data.terraform_remote_state.org in static context, which is required by local.account_ids
╵
╷
│ Error: Unable to compute static value
│ 
│   on providers.tf line 5, in provider "aws":
│    5:   for_each = toset(local.account_ids)
│ 
│ provider.aws.child.for_each depends on local.account_ids which is not available
╵
```

**Comment by apparentlymart** [comment]
Hi @kylan11,

Indeed, for this first iteration of the feature the `for_each` argument in `provider` blocks uses the same "early evaluation" mechanism that OpenTofu uses for expressions in `backend` blocks and module source strings. So the error you encountered is expected for the current version.

We are hoping to extend it to be fully-dynamic in future -- https://github.com/opentofu/opentofu/pull/2088 is representing that -- but we wanted to get an initial version of this out sooner, with less risk, and then iterate on it.

We'll be making new issues to represent some of the intentionally-omitted capabilities in this first release soon, as we get closer to the v1.9.0 beta period and its final release. I expect we'll add a comment linking to all of those issues here once we've created them, so that folks can vote on them separately to help prioritize that work.

For now you will need to specify your set of account ids either statically inside the configuration, or as an input variable.

Thanks!

**Comment by kylan11** [comment]
@apparentlymart Apologies for posting here! I read your last comment too late.
I've opened an issue for this use case.

Thank you!

**Comment by cam72cam** [comment]
Usecase1 is not covered by this issue, though there are other issues in the repo that discuss this problem.

You might be able to use a conditional for Usecase2, similar to how people use count to enable/disable resources.
Something like:
provider: `for_each = var.dns_record.provider == "cloudflare" ? {"enabled": true} : {}`
resource: `provider = cloudflare.alias["enabled"]`

I'm not 100% sure if that would work, but now I'm curious.

**Comment by lorengordon** [comment]
@hegerdes Use Case 1 is actually possible today, by creating a dependency between the EKS Cluster creation, and the helm/kubectl _providers_ and their authentication scheme. That way the providers wait to attempt authentication until after the cluster exists.

I worked it out a while back when using the terraform-aws-eks module and trying to clean up race conditions both on create and destroy workflows. Posted a note here: https://github.com/terraform-aws-modules/terraform-aws-eks/pull/3000#issuecomment-2059599312

**Comment by apparentlymart** [comment]
A variation of that DNS vendor use-case is what I wrote [the `terraformdns` modules](https://github.com/terraformdns) for. 

Those modules have a different architecture in mind, though: the modules that produce network services that should be registered in DNS each produce an output value containing a set of DNS record descriptions in a standardized shape, and then the _root_ module collects all of those (e.g. using `setunion`) and passes them to just _one_ of the vendor-specific modules, matching the vendor that is chosen for that particular configuration.

This therefore avoids every module needing to anticipate every possible DNS vendor, as long as each vendor has a provider with an API of a suitable shape to write a `terraformdns` module that wraps it. 

We were discussing some potential ways to make that pattern more economic recently over in https://github.com/opentofu/opentofu/issues/2151.
